rfcs/0182-taskcluster-yml-remote-references.md

```diff
@@ -0,0 +1,49 @@
+# RFC 182 - Allow remote references to .taskcluster.yml files processed by Taskcluster-GitHub
+* Comments: [#182](https://github.com/taskcluster/taskcluster-rfcs/pull/182)
+* Proposed by: @bhearsum
+
+# Summary
+
+Allow `.taskcluster.yml` files in GitHub repositories to outsource most of their contents to a `.taskcluster.yml` stored elsewhere.
+
+## Motivation
+
+`.taskcluster.yml` files in GitHub repositories are responsible for creating tasks (or not) in response to various GitHub events. In some cases (eg: when a decision task is used to delegate most of this to a separate tool), these files end up looking almost entirely identical, and hundreds of lines of JSON-e ends up duplicated across many different repositories. This duplication is a significant maintenance burden, and easily results in unwanted differences.
+
+Allowing a `.taskcluster.yml` to reference a file stored elsewhere will allow for deduplication of the bulk of the content in these files, resolving the concerns noted above.
+
+# Details
+
+Two new variables will be introduced to `.taskcluster.yml` files. The first, `config-from`, will allow an URL to be provided that points at a full-fledged `.taskcluster.yml` hosted elsewhere. The second, `context`, will allow for variables to be defined that will be included with the JSON-e context when the referenced `.taskcluster.yml` is rendered. Here is an example of a possible concrete `.taskcluster.yml`:
+
+```
+---
+version: 1
+config-from: https://github.com/taskcluster/taskgraph/blob/main/data/taskcluster-yml-github.yml
```

Comment:
Taking a page out of [GitHub's book](https://github.blog/2022-02-10-using-reusable-workflows-github-actions/#how-to-make-any-github-actions-workflow-reusable) which I like over the full url here would be a simpler format like:
`USER_OR_ORG_NAME/REPO_NAME/.github/workflows/REUSABLE_WORKFLOW_FILE.yml@TAG_OR_BRANCH`

I could see the full https url here useful if the user wants to use GitLab or Mercurial, for example. Maybe this `config-from` could be "smart" and visit the url, if provided, otherwise just use the GitHub shorthand method?

Reply:
I'm totally fine with using the format you suggest. Someone (Pete?) made the point before that this is a GitHub specific service, so using GitHub-specific references ought to be fine.

I mostly used full `https` links here to start with the most flexible thing, and have us pare down to something else if desired.

Reply:
We've previously tried to unify behavior across GH and Mercurial.  Maybe it makes sense to put `github.com` somewhere in the string, even if the result is not a URL that can be fetched?  This is what we've done for scope names, for example.

Reply:
> Does it make sense to pin specific version of the file? If that file is to be updated and we rerun event, then we could have completely different graphs. This is probably a feature, not a bug

Whether or not pinning is desirable is probably going to vary by project, and to be honest, I'm not even sure where we might land for Firefox CI projects. Obviously pinning gives you more reproducibility, but you lose some of the benefits of centralization if you have to manually go around and bump pins everywhere.

Perhaps it's best to allow for both pinned and unpinned, at least for now, until we have some better insight into how things work in practice?

> but maybe it would be good to somehow include this information in the decision task which version of the file was used. Something like attaching a hash of it as a metadata. 

I agree that we ought to include some information about what was used. `metadata` seems like a reason place for it. It might be nice to include revision as well, to make tracing things back easier. Another idea, probably not viable, would be to somehow pre-create an artifact for all tasks that end up being created, and attach this information there. If this were somehow possible, we could probably include the entire `.taskcluster.yml`, since we would no longer be limited to something reasonably sized.

> But then it would probably fail on rerun if this hash changes?

Hmmm, I'm not sure I fully understand what you're getting at. If I understand correctly, Taskcluster-GitHub would run once, create task definitions with this extra metadata, and then the tasks would eventually run. If they're rerun, the metadata would stay the same, even if the `.taskcluster.yml` had changed, since Taskcluster-GitHub is not involved with the reruns at all. (I think the only time we would get new metadata would be when we get new events, which would cause new task ids.)

Are you saying that the workers or queue or something else would be verifying the new information in the `metadata`? (I can certainly see ways reruns could fail in that case.)

Reply:
> Perhaps it's best to allow for both pinned and unpinned, at least for now, until we have some better insight into how things work in practice?

With either the URL or GH shorthand, we could reference a branch for unpinned and revision or tag for pinned. So we get both for free.

Reply:
> Obviously pinning gives you more reproducibility, but you lose some of the benefits of centralization if you have to manually go around and bump pins everywhere.
Agree with this and what Andrew says below. 
My thinking was if someone would be using "unofficial" source for the `config-from` then it would be a security concern if those tasks are not doing something unwanted.

> metadata seems like a reason place for it
Yes, this is what I was thinking of, too. We  should include few entries in metadatda by default (provided by github service at the time of parsing those configs) which files were used and probably their hashes (but maybe that's too much)

> if they're rerun, the metadata would stay the same, even if the .taskcluster.yml had changed, since Taskcluster-GitHub is not involved with the reruns at all.
For the individual tasks yes. I was thinking about the broader case, when you have a long living PR, and each push event would spawn a new decision task, so possibly a different set of tasks even. But this is rather an edgecase, and in this particular case it would be nicer to reference those `config-from` by revision. Not a problem I guess.

> Are you saying that the workers or queue or something else would be verifying the new information in the metadata?
no, probably doesn't make sense unless we would somehow need this for CoT? Just ignore for now :) Metadata is for humans to see what was the origin

Reply:
How do you folks feel about the latest version of this? Both with regard to the new proposed format for `config-from`, as well as the extra `context`?